internal/cli/server_upgrade.go

```diff
@@ -0,0 +1,362 @@
+	conn, err = serverclient.Connect(ctx,
+		serverclient.FromContextConfig(originalCfg),
+		serverclient.Timeout(5*time.Minute),
+	)
+	if err != nil {
+		c.ui.Output(
+			"Error connecting to server: %s\n\n%s",
+			clierrors.Humanize(err),
+			errInstallRunning,
+			terminal.WithErrorStyle(),
+		)
+		return 1
+	}
+	client = pb.NewWaypointClient(conn)
+
+	resp, err = client.GetVersionInfo(ctx, &empty.Empty{})
+	if err != nil {
+		c.ui.Output(
+			"Error retrieving server version info: %s", clierrors.Humanize(err),
+			terminal.WithErrorStyle())
+		return 1
+	}
+
+	c.ui.Output("\nServer upgrade for platform %q context %q complete!",
+		c.platform, ctxName, terminal.WithSuccessStyle())
+
+	c.ui.Output("Waypoint has finished upgrading the server to version %q\n",
+		resp.Info.Version, terminal.WithSuccessStyle())
+
+	c.ui.Output(addrSuccess, advertiseAddr.Addr, "https://"+httpAddr,
+		terminal.WithSuccessStyle())
+
+	return 0
+}
+
+func (c *ServerUpgradeCommand) Flags() *flag.Sets {
+	return c.flagSet(0, func(set *flag.Sets) {
+		f := set.NewSet("Command Options")
+		f.BoolVar(&flag.BoolVar{
+			Name:    "auto-approve",
+			Target:  &c.confirm,
+			Default: false,
+			Usage:   "Confirm server upgrade.",
+		})
+		f.StringVar(&flag.StringVar{
+			Name:    "context-name",
+			Target:  &c.contextName,
+			Default: "",
+			Usage:   "Waypoint server context to upgrade.",
+		})
+		f.StringVar(&flag.StringVar{
+			Name:    "platform",
+			Target:  &c.platform,
+			Default: "",
+			Usage:   "Platform to upgrade the Waypoint server from.",
+		})
+		f.StringVar(&flag.StringVar{
+			Name:    "snapshot-name",
+			Target:  &c.snapshotName,
+			Default: "",
+			Usage:   "Platform to upgrade the Waypoint server from.",
+		})
+		f.BoolVar(&flag.BoolVar{
+			Name:    "skip-snapshot",
+			Target:  &c.skipSnapshot,
+			Default: false,
+			Usage:   "Skip creating a snapshot of the Waypoint server.",
+		})
+
+		for name, platform := range serverinstall.Platforms {
+			platformSet := set.NewSet(name + " Options")
+			platform.UpgradeFlags(platformSet)
+		}
+	})
+}
+
+func (c *ServerUpgradeCommand) AutocompleteArgs() complete.Predictor {
+	return complete.PredictNothing
+}
+
+func (c *ServerUpgradeCommand) AutocompleteFlags() complete.Flags {
+	return c.Flags().Completions()
+}
+
+func (c *ServerUpgradeCommand) Synopsis() string {
+	return "Upgrades Waypoint server in the current context to the latest version"
+}
+
+func (c *ServerUpgradeCommand) Help() string {
+	return formatHelp(`
+Usage: waypoint server upgrade [options]
+
+	Upgrade Waypoint server in the current context to the latest version or the
+	server image version specified. By default, Waypoint will upgrade to server
+	version "hashicorp/waypoint:latest". Before upgrading, a snapshot of the
+	server will be taken in case of any upgrade failures.
+
+` + c.Flags().Help())
+}
+
+var (
+	defaultSnapshotName = "waypoint-server-snapshot"
```

Comment:
Is there any reason not to have this be the default on the `snapshot-name` flag? Setting this directly as a default on the flag can clear up some of the logic on lines 157-159.

Reply:
Ah, yes! I was thinking if we left a default, we wouldn't get the timestamp and we'd overwrite previous snapshots with the same default name. But maybe our default can just be the "base name" and it operates the same as this in the case of no snapshot filename specified?

Reply:
Hm, I would rather not overwrite old snapshots automatically. I understand the point of that is to prevent someone from stacking up old snapshots, but what if they try an upgrade, it borks, they try again, and the second one also borks but they are left with a weird snapshot state now?

Reply:
Ah, I think I wasn't clear enough in my previous comment! The way it works now is it doesn't overwrite previous snapshots, it uses this as a base name and appends a timestamp, so each run is a new snapshot.

I meant to say if we had set a default for the CLI, the help text would give the impression that it would save as that filename exactly, when in reality it shouldn't because we'd overwrite previous snapshots, so that's why the default value in the snapshot flag was left blank.

But maybe we set a default for the flag as the "basename" and still apply the timestamp at the end? Then update the help text to explain a timestamp is going to be appended to the default name.